engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java

```diff
@@ -908,7 +908,7 @@ private void execute() throws YarnException, IOException
           LOG.info("Releasing {} as resource with priority {} was already assigned", allocatedContainer.getId(), allocatedContainer.getPriority());
           releasedContainers.add(allocatedContainer.getId());
           numReleasedContainers++;
-          numRequestedContainers--;
+          numRequestedContainers++;    // undo the decrement above for this allocated container
```

Comment:
@vrozov I traced the scenario you described and it is indeed problematic (but a corner case and an existing bug). Basically in your scenario a container is requested twice (because the first request times out) and that same container also becomes redundant (because the AM decides it doesn't need it anymore). So the way the logic works when Yarn returns 2 containers is the code processes the first allocated container and because it is obviously not "already allocated" it does not touch the counters and then removes the outstanding request from requestedResources Map. But later it also figures out the container is not needed anymore so it creates a release-request for that container. While processing the second allocated-container it cannot determine this to be "already allocated" case because the request was removed requestedResources and so numRequestedContainers doesn't get incremented. It again figures out this container is not needed so creates a release-request for this container. But then numRequestedContainers stays at -1 and that is a problem.